text/0039-reduce-module-size.md

```diff
@@ -0,0 +1,550 @@
+breaking change for customers who are using these Lambda Layers in
+network-restricted environments. If customers are able to use the CDK CLI in
+their network-restricted environments, then it is reasonable to assume they were
+able to acquire the `aws-cdk` package from npm somehow. This implies customers
+will be able to also acquire the `@aws-cdk/lambda-layer-X` packages from npm. They can either
+make sure the necessary packages are already installed and available in the
+`node_modules` directory where their CDK applications are executed, or include
+them in a private npm registry.
+
+Unfortunately, there is no way to remove these large files and host them
+somewhere else without causing a breaking change for these customers. We will
+publish a CLI notice directly to potentially-affected customers of this change
+well in advance of releasing it. And, advertise the change in advance of release
+on the aws-cdk GitHub repository.
+
+##### What about customers in ADC regions?
+
+Today, there is a mechanism for publishing AWS CDK into ADC (Amazon Dedicated
+Cloud) regions. This mechanism includes bundling the direct dependencies of the
+CDK and publishing them into the ADC regions, e.g. `constructs`. We will add the
+`@aws-cdk/lambda-layer-X` packages to this process, and when customers use the aws-cdk npm
+package in ADC regions, the right dependencies will already be present in the
+`node_modules` directory.
+
+#### .jsii Files
+
+There are two files, .jsii.tabl.json and .jsii, bundled in `aws-cdk-lib` for the
+JSII runtime to work for any non-NodeJS languages. And, to enable code and
+documentation generation from the published npm package. Today, they make up 42%
+of the package size. And in the future, after adding more Lambda Layers, the .jsii files would
+compose 30% of the package size.
+
+We will compress the .jsii.tabl.json and .jsii files. Compressing with gzip
+currently creates a 4 MB and 4.7 MB file, respectively. This change will impact
+the JSII runtime, and Construct Hub. NodeJS JSII runtimes don't need to use
+either file, and non-NodeJS runtimes would decompress the .jsii file when
+needed. Construct Hub generates documentation from the .jsii.tabl.json file, and
+it will also decompress this file as needed.
+
+#### Source Map Files (.js.map)
+
+These files currently makeup 22% of the total size of `aws-cdk-lib`. Source map
+files map from the transformed source to the original source. In the case of
+`aws-cdk-lib`, these are not actually useful today, since we do not ship the
+original `.ts` files. We will remove the source maps from the released package.
+
+#### Javascript Files (.js)
+
+The Javascript files make up about 10% of the overall package size. Today, they
+are already minified in the released package. There is not much else that can be
+done to make them smaller.
+
+#### Type Declaration Files (.d.ts)
+
+We can minify these files, which could result in a space saving of up to 20 MB.
```

Comment:
This is a tempting number, but people command-clicking to CDK definitions will end up in a file with unreadable garbage. Not sure this would be worth it.

I wonder if there are gains to be had if we combine the `.d.ts` files for submodules together (into 1 file per submodule). I'm guessing this would lead to better compression, though the difference might not be very significant. If nothing else,  would have performance gains as well for downstream `tsc` invocations, since the compiler wouldn't have to read so many files.

Reply:
Oh, I thought we only needed to worry about API reference in IDEs. But yes this will be an issue as well.